lib/ExtraCorres.thy

```diff
@@ -241,17 +241,19 @@ lemma hoare_from_abs_inv:
 lemma in_whileLoop_corres:
   assumes body_corres:
     "\<And>r r'. rrel r r' \<Longrightarrow>
-             corres_underlying srel False nf' rrel (P and C r) (P' and C' r') (B r) (B' r')"
-  and body_inv: "\<And>r. \<lbrace>P and C r\<rbrace> B r \<lbrace>\<lambda>_. P\<rbrace>"
-                "\<And>r'. \<lbrace>P' and C' r'\<rbrace> B' r' \<lbrace>\<lambda>_. P'\<rbrace>"
-  and cond: "\<And>r r' s s'. \<lbrakk>rrel r r'; (s, s') \<in> srel; P s; P' s'\<rbrakk> \<Longrightarrow> C r s = C' r' s'"
+             corres_underlying srel nf nf' rrel (P r and C r) (P' r' and C' r') (B r) (B' r')"
+  and body_inv: "\<And>r. \<lbrace>P r and C r\<rbrace> B r \<lbrace>\<lambda>r. P r\<rbrace>"
+                "\<And>r'. \<lbrace>P' r' and C' r'\<rbrace> B' r' \<lbrace>\<lambda>r'. P' r'\<rbrace>"
```

Comment:
I'm undecided myself, but could using `rv` in these postconditions to distinguish from the bound `r` improve readability? I guess a potential concern is that it could also be confused with the final return value of the whole loop.
```suggestion
  and body_inv: "\<And>r. \<lbrace>P r and C r\<rbrace> B r \<lbrace>\<lambda>rv. P rv\<rbrace>"
                "\<And>r'. \<lbrace>P' r' and C' r'\<rbrace> B' r' \<lbrace>\<lambda>rv'. P' rv'\<rbrace>"
```

Reply:
The situation is quite tricky with these `whileLoop`s. There is the (blue) initial value given to the `whileLoop` (you can see an example of this in the conclusion of the statement of `corres_whileLoop_ret`). Then there is the (green) return value given by the loop body. I think it would be nicer to stick with the `r` for here, to reinforce that connection. But happy to discuss it some more

Reply:
The other alternative for these `valid` bits is that I could just eta contract the postcondition. I left it written out to emphasise that these guards do talk about the return value, but maybe that's clear enough

Reply:
I would usually prefer the eta-contracted version unless we're trying to do the bound-name preservation trick where Isabelle would replace `rv` with whatever name was in the goal (which is not the case here).